### ISSUE TITLE
adding + for strictly positive data type arguments

### ISSUE THREAD

**Issue by mariari** [anoma/juvix#404]
There was a suggestion that we add `+` before types to mean an argument to a data type is strictly positive. With some syntax like

```haskell
data Fun (+ A) ...
```

What this gains, is to force A to be strictly positive in Fun.

 The main questions I think before we add it to the frontend syntax is

1) is strict positivity always inferable
    - If this is not the case we lose the ability to describe things!
    - This was brought up, as if it isn't, then you won't be able to infer args as strictly positive
    - Meaning we lose expressiveness to abstract over say a rose tree. 

2) what dangerous behavior does this pervent
    - In this case, we lose some saftey through not enough restrictions!
    - Is there some use case where a thing is forced to be non-strictly positve ever create any harmful patterns

One issue with bringing in more syntax, is that when a new user sees

```haskell
data Fun (+ A) ...
```

they will have to search for it (kind of hard to search), and then figure out what `posititvity` means. This is not something a user would normally think about, as one doesn't typically think about positivity when constructing types. Thus they will have to care about a detail that may not matter

if 1) and the set of things in 2) is empty, then I think there is little benefit gained for adding this
    - Really just confusing new users when they see it in a codebase or forced to use it

However if that is not the case, then there is significant enough reason to add it to the front end language.

**Comment by mariari** [comment]
Another thing one should prove before dismissing `+` here, is will the compilation time increase in any meaningful way.

I think in this case it does not, and it should be rather fast.

We should always consider the time of compilation when it comes to inference

**Comment by mariari** [comment]
From a discussion with Andy:

We determined this problem can be broken down into refinement types


```idris

data Fun A : A {Positive A} -> Fun = 
...
```

with an arrow defined 

```idris
A +-> Type = A: Type {positive A} → Type
```

```idris
data Fun A : A -+> Fun = 
```

If a type isn't positive, then the set is empty and this is invalid.

This is a technique already used by the user, and so should not encur any cost.


Another point brought up is that we have to denote positivity in explicit module signatures, as one has to know if it is positive or not.

**Comment by thealmarty** [comment]
Without this feature (of letting users declare positive parameters), a user cannot ask the type checker to make sure a parameter is strictly positive. When you don't add the +, the type checker works on the assumption that it isn't a requirement for it to be strictly positive. You won't get an error. So it should not confuse the user.

I think the original proposal has a much clearer syntax than proposed above.

I can make a PR and make the type checker infer whether the parameters are strictly positive or not, and make it so that the type checker works on the inferred results. (It means in `sposConstructor` the list `[pos]` would be inferred instead of given by the user) This should solve the problem of some data types being rejected even though they may be fine. 

However there are 2 drawbacks:

- because I have to check strict positivity of the parameters for all the constructors, it will be quite cumbersome. 
- the user may not as easily figure out why a data type is rejected if the user made a mistake in the data declaration. The error message cannot be as targeted.

In short, I recommend having this feature because it lets the users specify the data type more precisely, which is the same reason we want to allow dependent types and first class usages.

**Comment by mariari** [comment]
> I think the original proposal has a much clearer syntax than proposed above.

I think the main issue here is that we are introducing special cased syntax the user has to memorize and learn how to search (searching for data types with `+` in scala, is a pain to find what it means and where you should use it. and ask about.

However if a new user comes across (or the compiler hands them) code with the refinement, then it'll look just like anything else. 

Even the `+->` arrow alias (or whatever signifier we would use) will be desugared by some tool we have (we have to do this anyways, to show users the `->` usage).

Thus no new forms will have to be added.

> When you don't add the +, the type checker works on the assumption that it isn't a requirement for it to be strictly positive. You won't get an error. So it should not confuse the user.

Yes in those cases the user shouldn't be confused, however when a user comes across this in someone elses code then there will be an issue.

Often codebases aren't written by 1 person, and when we introduce more forms to a language, that just adds a barrier of entry to a codebase.

Often this is considered the complexity budget, and adding any special forms to the compiler is doing this (OCaml's module syntax suffers from this, scala is a good case of this, etc.).

> However there are 2 drawbacks:

> because I have to check strict positivity of the parameters for all the constructors, it will be quite 
   cumbersome.


>  the user may not as easily figure out why a data type is rejected if the user made a mistake in the 
  data declaration. The error message cannot be as targeted.

As per the first point, by the time it hits there it may be infered already and de-sugared or it would be another pass so thus an independent isolated pass.

As per the second point, we can easily direct it, as we can say the type expects a positive typed argument, and the one given isn't. We can even mention that the positivity is inferred.

**Comment by mariari** [comment]
About annotations/Inferrence:

https://www.researchgate.net/publication/269080460_A_hierarchy_of_mendler_style_recursion_combinators

It seems in this paper they do not annotate when negative nor positive occurences happen, they just have the expressions for free

**Comment by cwgoes** [comment]
Personally I find the arrow syntax less clear than just putting a `+` before the type. I don't quite understand how we would represent this as a refinement - positivity isn't in the *type theory* - are you proposing that we add it? Or is it just syntactic sugar?

Generally I am in favour of inferring positivity by default if we can do so safely, but also allowing the user to provide explicit annotations, even if the inference is excellent. However, I don't think inference is worth implementing right now, this just isn't that high-priority (definitely worth noting though).

> If we have language extensions ala Haskell, then this is fine, afterall in haskell we have extensions to say

Whether or not to have language extensions is a larger question. I do think there is a real complexity cost here - certainly most of Haskell's type-theoretic extensions we should be able to avoid the need for. Syntactic ones I am less sure about; I think it might be preferable to allow user-defined macros or `syntax` definitions (a la Idris) instead of having lots of them built into the compiler itself (but if users have to write lots of macros, we've probably done something wrong).

**Comment by A-Manning** [comment]
> This was brought up, as if it isn't, then you won't be able to infer args as strictly positive
Meaning we lose expressiveness to abstract over say a rose tree.

It's entirely possible to express rose trees with strict positivity, provided that you have induction-induction (aka. mutual induction). Here is an example in F*:

```fstar
type t (a : Type u#a): Type u#a =
  | Node: a -> branches a -> t a

and branches (a : Type u#a): Type u#a =
  | Branches: list (t a) -> branches a
```

**Comment by andy-morris** [comment]
Sure, the problem is that you can't write one that's generalised over the list type without being able to say something like

```agda
data Rose′ (F : + Set → Set) (A : Set) : Set where
  rose : A → F (Rose′ F A) → Rose′ F A
```

(But if anyone can think of a less contrived example than this that would probably be a good thing to have...)

**Comment by cwgoes** [comment]
Per discussion with @mariari we agreed that using refinement syntax is not prudent because positivity checking doesn't behave like a refinement, so it would be confusing - inference would be nice in the future, the explicit annotation seems fine for now, maybe we could use a longer keyword (`positive`) instead of just `+` for easier searching.

**Comment by cwgoes** [comment]
Still to be considered but open questions regarding parsability.

**Comment by Acentelles** [comment]
Closing in favor of https://hackmd.io/minsE9gHQ0CMv0x3UeY2Uw
